Mathlib/Algebra/MonoidAlgebra/Star.lean

```diff
@@ -0,0 +1,174 @@
+  map_mul' := star_mul'
+  map_add' := star_add
+  commutes' := by
+    letI : StarRing k := starRingOfComm
+    have : TrivialStar k := ⟨congrFun rfl⟩
+    simp [Algebra.algebraMap_eq_smul_one]
+
+@[simp] lemma invert_apply {f : MonoidAlgebra k G} {x : G} : invert f x = f x⁻¹ := rfl
+
+lemma involutive_invert : Function.Involutive (invert (k := k) (G := G)) := star_star
+
+end Commutative
+
+end MonoidAlgebra
+
+namespace AddMonoidAlgebra
+
+section
+
+variable {k G : Type _} [Semiring k] [StarRing k] [DecidableEq G] [AddGroup G]
+
+/-- A `StarRing` structure on the additive group algebra.
+
+See `MonoidAlgebra.instStarRing` for further remarks. -/
+def instStarRing : StarRing (AddMonoidAlgebra k G) where
+  star f := ⟨f.support.map (Equiv.neg G), star ∘ f ∘ Neg.neg, fun x ↦ by
+    simp only [Finset.mem_map_equiv, Finsupp.mem_support_iff, Function.comp_apply, star_ne_zero,
+      Equiv.neg_symm, Equiv.neg_apply]⟩
+  star_involutive := fun f ↦ Finsupp.ext fun x ↦ by simp
+  star_mul := fun f g ↦ Finsupp.ext fun x ↦ by
+    simp only [Finsupp.coe_mk, Function.comp_apply]
+    rw [AddMonoidAlgebra.mul_apply, AddMonoidAlgebra.mul_apply]
+    simp only [Finsupp.sum, star_sum, apply_ite, star_mul, star_zero, Finsupp.coe_mk,
+      Function.comp_apply, Finset.sum_map, Equiv.coe_toEmbedding, Equiv.neg_apply, neg_neg]
+    rw [Finset.sum_comm]
+    simp_rw [← neg_add_rev, neg_eq_iff_eq_neg]
+  star_add := fun f g ↦ Finsupp.ext fun x ↦ by
+    simp only [Finsupp.coe_mk, Function.comp_apply]
+    rw [Finsupp.add_apply, Finsupp.add_apply]
+    simp
+
+attribute [local instance] instStarRing
+
+@[simp] lemma star_apply {f : AddMonoidAlgebra k G} {x : G} : star f x = star (f (-x)) := rfl
+
+@[simp] lemma star_smul {a : k} {f : AddMonoidAlgebra k G} :
+    star (a • f) = (MulOpposite.op (star a)) • star f := Finsupp.ext fun x ↦ by
+  rw [Finsupp.smul_apply, star_apply, Finsupp.smul_apply, smul_eq_mul, star_mul, star_apply,
+    MulOpposite.smul_eq_mul_unop, MulOpposite.unop_op]
+
+end
+
+section Commutative
+
+variable {k G : Type*} [CommSemiring k] [DecidableEq G] [AddCommGroup G]
+
+/-- When both `k` and `G` are commutative, we do not need a star on `k` to obtain a `StarRing`
+structure. See `AddMonoidAlgebra.invert` where this is bundled as an algebra equivalence. -/
+def instStarRingOfCommComm {k : Type*} [CommSemiring k] : StarRing (AddMonoidAlgebra k G) :=
+  letI : StarRing k := starRingOfComm; AddMonoidAlgebra.instStarRing
+
+attribute [local instance] instStarRingOfCommComm
+
+/-- The natural inversion carried by the group algebra of an additive commutative group over
+commutative scalars. -/
+def invert : AddMonoidAlgebra k G ≃ₐ[k] AddMonoidAlgebra k G where
+  toFun := star
+  invFun := star
+  left_inv := star_star
+  right_inv := star_star
+  map_mul' := star_mul'
+  map_add' := star_add
+  commutes' := by
+    letI : StarRing k := starRingOfComm
+    have : TrivialStar k := ⟨congrFun rfl⟩
+    simp [Algebra.algebraMap_eq_smul_one]
```

Comment:
I'd argue it would be clearer to build this directly using `mapDomainAlgHom` with `MulEquiv.Inv`

Reply:
Right, agreed. OK attempt number three.